### PR TITLE
Add CI/CD for Docker Build and Sonar cloud

### PR DESCRIPTION
**Pull request by SaikrishnaBairamoni**
This PR is to add the GitHub actions workflows for both docker hub build and to run sonar scan analysis for publishing static reports on sonar cloud dashboard.

### CODE REVIEW

**Review by codygarver — CHANGES_REQUESTED** [review]
sonar-scanner.properties is not being used by the sonar-scanner in ci.yml

ci.yml is also needs to include [tmca-tester](https://github.com/usdot-jpo-ode/jpo-security-svcs/tree/develop/tmca-tester)

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+         path: /__w/jpo-security-svcs/jpo-security-svcs/target 
+         if-no-files-found: error
+     - name: build tmca-tester
+       uses: actions/setup-node@v3
+       with:
+         node-version: '18.x'
+     - run: |
+         cd /__w/jpo-security-svcs/jpo-security-svcs/tmca-tester
+         npm install node-menu@1.3.2
+         npm i
```
Comment (codygarver) [inline .github/workflows/ci.yml]: `npm i` is unnecessary since the next line does npm install

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+         node ./tmca-tester.js  
+  sonar:
+    needs: [jpo-security-svcs]
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout ${{ github.event.repository.name }}
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0   
+      - run: chmod +x -R /home/runner/work/jpo-security-svcs/jpo-security-svcs
```
Comment (codygarver) [inline .github/workflows/ci.yml]: Does the build work without this?

File: .github/workflows/docker.yml [hunk 3]
```diff
@@ -0,0 +1,24 @@
+name: Docker build
+
+on:
+  push:
+    branches-ignore:
+      - "develop"
+      - "release/*"
```
Comment (codygarver) [inline .github/workflows/docker.yml]: Do we want `master` in this list?

File: .github/workflows/docker.yml [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+      - run: ls -la && pwd 
+      - name: Build
+        uses: docker/build-push-action@v3
+        with:
+          context: ./
+          file: ./Dockerfile
```
Comment (codygarver) [inline .github/workflows/docker.yml]: I think these lines are unnecessary and can be removed

File: .github/workflows/dockerhub.yml [hunk 5]
```diff
@@ -0,0 +1,25 @@
+name: "DockerHub Build and Push"
+
+on:
+  push:
+    branches:
+      - "develop"
```
Comment (codygarver) [inline .github/workflows/dockerhub.yml:6]: Do we want to also include `master` and `release/*`

File: .github/workflows/dockerhub.yml [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+jobs:
+  dockerhub-Jpo-ode:
```
Comment (codygarver) [inline .github/workflows/dockerhub.yml]: lowercase this `J`

File: .sonarqube/sonar-scanner.properties [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+#sonar.sourceEncoding=UTF-8
+
+
+sonar.modules=jpo-security-svcs
+
+jpo-security-svcs.sonar.projectBaseDir          = /home/runner/work/jpo-security-svcs/jpo-security-svcs
+
+jpo-security-svcs.sonar.sources         = src
+
+#jpo-security-svcs.sonar.tests                = src/test
```
Comment (codygarver) [inline .sonarqube/sonar-scanner.properties]: Delete this whole file, it doesn't get used anywhere

File: pom.xml [hunk 8]
```diff
@@ -18,6 +16,20 @@
...
+		<jmockit.version>1.49</jmockit.version>
+		<argLine>-javaagent:${user.home}/.m2/repository/org/jmockit/jmockit/${jmockit.version}/jmockit-${jmockit.version}.jar</argLine>
+		<sonar.organization>usdot-jpo-ode-1</sonar.organization>
+        <sonar.host.url>https://sonarcloud.io</sonar.host.url>
+        <!-- JaCoCo Properties -->
+        <jacoco.version>0.8.8</jacoco.version>
+        <sonar.java.coveragePlugin>jacoco</sonar.java.coveragePlugin>
+        <sonar.dynamicAnalysis>reuseReports</sonar.dynamicAnalysis>
+        <sonar.coverage.jacoco.xmlReportPaths>${project.basedir}/target/site/jacoco/jacoco.xml</sonar.coverage.jacoco.xmlReportPaths>
+        <sonar.language>java</sonar.language>
```
Comment (codygarver) [inline pom.xml]: Replace tabs with spaces and properly indent within `<properties>`

File: pom.xml [hunk 9]
```diff
@@ -64,21 +76,84 @@
...
-				<groupId>org.springframework.boot</groupId>
-				<artifactId>spring-boot-maven-plugin</artifactId>
-			</plugin>
+         <plugin>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-maven-plugin</artifactId>
+         </plugin>
+         <plugin>
+          <groupId>org.apache.maven.plugins</groupId>
+          <artifactId>maven-surefire-plugin</artifactId>
```
Comment (codygarver) [inline pom.xml]: When does anything with `surefire` get used?

File: pom.xml [hunk 10]
```diff
@@ -64,21 +76,84 @@
...
+                <enableProcessChecker>all</enableProcessChecker>
+                <shutdown>exit</shutdown>
+            <systemPropertyVariables>
+              <loader.path>${loader.path}</loader.path>
+              <buildDirectory>${project.build.directory}</buildDirectory>
+            </systemPropertyVariables>
+          </configuration>
+          <dependencies>
+            <dependency>
+              <groupId>org.apache.maven.surefire</groupId>
```
Comment (codygarver) [inline pom.xml]: When does anything with `surefire` get used?

**Review by codygarver — CHANGES_REQUESTED** [review]
See diff comments

File: .github/workflows/ci.yml (hunk 1 above)
Comment (SaikrishnaBairamoni) [reply .github/workflows/ci.yml]: removed

File: .github/workflows/docker.yml (hunk 3 above)
Comment (SaikrishnaBairamoni) [reply .github/workflows/docker.yml]: updated

File: .github/workflows/dockerhub.yml (hunk 5 above)
Comment (SaikrishnaBairamoni) [reply .github/workflows/dockerhub.yml:6]: updated

File: .sonarqube/sonar-scanner.properties (hunk 7 above)
Comment (SaikrishnaBairamoni) [reply .sonarqube/sonar-scanner.properties]: deleted

File: .github/workflows/docker.yml [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+      
+jobs:
+  jpo-security-svcs:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+      - run: ls -la && pwd
```
Comment (codygarver) [inline .github/workflows/docker.yml:19]: Remove this `run:` line

File: pom.xml [hunk 12]
```diff
@@ -18,6 +16,20 @@
 	<packaging>jar</packaging>
 	<name>jpo-security-svcs</name>
 	<description>JPO ODE Cryptography Module</description>
+	<properties>
+		      <java.version>1.8</java.version>
+		      <jmockit.version>1.49</jmockit.version>
+		      <argLine>-javaagent:${user.home}/.m2/repository/org/jmockit/jmockit/${jmockit.version}/jmockit-${jmockit.version}.jar</argLine>
+		      <sonar.organization>usdot-jpo-ode-1</sonar.organization>
+          <sonar.host.url>https://sonarcloud.io</sonar.host.url>
+          <!-- JaCoCo Properties -->
```
Comment (codygarver) [inline pom.xml]: Fix this alignment, too many spaces

**Review by codygarver — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+         name: jpo-security-svcs
+         path: /__w/jpo-security-svcs/jpo-security-svcs/target 
+         if-no-files-found: error
+     - name: build tmca-tester
+       uses: actions/setup-node@v3
+       with:
+         node-version: '18.x'
+     - run: |
+         cd /__w/jpo-security-svcs/jpo-security-svcs/tmca-tester
+         npm install node-menu@1.3.2
```
Comment (codygarver) [inline .github/workflows/ci.yml]: Remove this line about node-menu, no need to manually install a dependency already declared in package.json

File: .github/workflows/ci.yml [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+         ls -ls && pwd
+         mvn -e -X clean org.jacoco:jacoco-maven-plugin:prepare-agent package
+         mvn -e -X clean org.jacoco:jacoco-maven-plugin:report package
+     - name: Archive code coverage results
+       uses: actions/upload-artifact@v3
+       with:
+         name: jpo-security-svcs
+         path: /__w/jpo-security-svcs/jpo-security-svcs/target 
+         if-no-files-found: error
+     - name: build tmca-tester
```
Comment (codygarver) [inline .github/workflows/ci.yml:40]: tmca-tester should be a dedicated job so it can run in parallel

File: .github/workflows/ci.yml [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+       with:
+         node-version: '18.x'
+     - run: |
+         cd /__w/jpo-security-svcs/jpo-security-svcs/tmca-tester
+         npm install node-menu@1.3.2
+         npm install --force
+         npm run build --if-present      
+         node ./tmca-tester.js  
+  sonar:
+    needs: [jpo-security-svcs]
```
Comment (codygarver) [inline .github/workflows/ci.yml]: Add tmca-tester job as a dependency when you split it to its own job

**Review by codygarver — CHANGES_REQUESTED** [review]
See diff in comments

File: .github/workflows/ci.yml [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+     - name: Checkout ${{ github.event.repository.name }}
+       uses: actions/checkout@v3
+     - name: build tmca-tester
+       uses: actions/setup-node@v3
+       with:
+         node-version: '18.x'
+     - run: |
+         cd $GITHUB_WORKSPACE/tmca-tester
+         npm install --force
+         npm run build --if-present
```
Comment (codygarver) [inline .github/workflows/ci.yml]: Why `--if-present`?

**Review by codygarver — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 16 above)
Comment (SaikrishnaBairamoni) [reply .github/workflows/ci.yml]: removed

**Review by codygarver — APPROVED** [review]

**Pull request merged**